lib/index.ts

```diff
@@ -0,0 +1,173 @@
+		deviceType: string,
+		currentVersion: string,
+		targetVersion: string,
+	) {
+		if (!rSemver.valid(currentVersion)) {
+			throw new Error('Invalid current balenaOS version');
+		}
+
+		if (!rSemver.valid(targetVersion)) {
+			throw new Error('Invalid target balenaOS version');
+		}
+
+		if (
+			rSemver.prerelease(currentVersion) ||
+			rSemver.prerelease(targetVersion)
+		) {
+			throw new Error(
+				'Updates cannot be performed on pre-release balenaOS versions',
+			);
+		}
+
+		if (isDevVariant(currentVersion) || isDevVariant(targetVersion)) {
+			throw new Error(
+				'Updates cannot be performed on development balenaOS variants',
+			);
+		}
+
+		if (rSemver.lt(targetVersion, currentVersion)) {
+			throw new Error('OS downgrades are not allowed');
+		}
+
+		if (rSemver.compare(currentVersion, targetVersion) === 0) {
+			throw new Error('Current OS version matches Target OS version');
+		}
+
+		const fromMajor = rSemver.major(currentVersion);
+		const toMajor = rSemver.major(targetVersion);
+		const actionName = `resinhup${fromMajor}${toMajor}` as ActionName;
+
+		const { actionsConfig } = this;
+		const deviceSpecific = actionsConfig.deviceTypes[deviceType];
+
+		if (deviceSpecific == null || deviceSpecific[actionName] == null) {
+			throw new Error(
+				`This update request cannot be performed on '${deviceType}'`,
+			);
+		}
+
+		const {
+			minSourceVersion,
+			targetMajorVersion,
+			minTargetVersion,
+			maxTargetVersion,
+		} = {
+			...actionsConfig.actions[actionName],
+			...deviceSpecific[actionName],
+		};
+
+		if (rSemver.lt(currentVersion, minSourceVersion)) {
+			throw new Error(`Current OS version must be >= ${minSourceVersion}`);
+		}
+
+		if (rSemver.major(targetVersion) !== targetMajorVersion) {
+			throw new Error(
+				`Target OS version must be of major version ${targetMajorVersion}`,
+			);
+		}
+
+		if (rSemver.lt(targetVersion, minTargetVersion)) {
+			throw new Error(`Target OS version must be >= ${minTargetVersion}`);
+		}
+
+		if (maxTargetVersion && rSemver.gte(targetVersion, maxTargetVersion!)) {
+			throw new Error(`Target OS version must be < ${maxTargetVersion}`);
+		}
+
+		return actionName;
+	}
+
+	/**
+	 * @summary Returns whether the provided device type supports OS updates between the current and target balenaOS versions
+	 * @name isSupportedOsUpdate
+	 * @public
+	 * @function
+	 * @memberof HUPActionHelper
+	 *
+	 * @param {String} deviceType - device type slug
+	 * @param {String} currentVersion - the current semver balenaOS version on the device
+	 * @param {String} targetVersion - the target semver balenaOS version
+	 *
+	 * @returns {Boolean}
+	 *
+	 * @example
+	 * hupActionHelper.isSupportedOsUpdate('raspberrypi3', '2.0.0+rev1.prod', '2.2.0+rev1.prod');
+	 */
+	isSupportedOsUpdate(
+		deviceType: string,
+		currentVersion: string,
+		targetVersion: string,
+	) {
+		try {
+			return !!this.getHUPActionType(deviceType, currentVersion, targetVersion);
+		} catch (err) {
+			return false;
```

Comment:
I think it's better not to swallow exceptions if they are not expected. Why don't we just let the error propagate if not expected, and handle the other cases accordingly?

Reply:
My reasoning was that if we really care about the exception & its message then we should go for `getHUPActionType`. The exceptions thrown afaict are just a way to encapsulate the error messages. This method was indeed intending to only return a boolean so that it can be used in `if` statement & `[].filter`.
I would prefer to have such a method for the above reasons.